Objects/unicodeobject.c

```diff
@@ -2272,34 +2269,85 @@ PyUnicode_FromString(const char *u)
     return PyUnicode_DecodeUTF8Stateful(u, (Py_ssize_t)size, NULL, NULL);
 }
 
+
 PyObject *
 _PyUnicode_FromId(_Py_Identifier *id)
 {
-    if (!id->object) {
-        id->object = PyUnicode_DecodeUTF8Stateful(id->string,
-                                                  strlen(id->string),
-                                                  NULL, NULL);
-        if (!id->object)
+    PyInterpreterState *interp = _PyInterpreterState_GET();
+    struct _Py_unicode_ids *ids = &interp->unicode.ids;
+
+    // Copy the index since _Py_Identifier.index is declared as volatile
+    Py_ssize_t index = id->index;
```

Comment:
Since reading the index is not guarded by lock, it is possible that we read index simultaneously with writing it in other thread. In such case the half of index bits can be old, and the other half is new. We need not just add `volatile` for index, but use an atomic integer instead of just `Py_ssize_t`.